plugins/module_utils/s3.py

```diff
@@ -100,3 +102,63 @@ def validate_bucket_name(module, name):
     if name[-1] not in string.ascii_lowercase + string.digits:
         module.fail_json(msg='bucket names must begin and end with a letter or number')
     return True
+
+
+# Spot special case of fakes3.
+def is_fakes3(url):
+    """ Return True if endpoint_url has scheme fakes3:// """
+    result = False
+    if url is not None:
+        result = urlparse(url).scheme in ('fakes3', 'fakes3s')
+    return result
+
+
+def parse_fakes3_endpoint(url):
+    fakes3 = urlparse(url)
+    protocol = "http"
+    port = fakes3.port or 80
+    if fakes3.scheme == 'fakes3s':
+        protocol = "https"
+        port = fakes3.port or 443
+    endpoint_url = f"{protocol}://{fakes3.hostname}:{to_text(port)}"
+    use_ssl = bool(fakes3.scheme == 'fakes3s')
+    return {"endpoint": endpoint_url, "use_ssl": use_ssl}
+
+
+def parse_ceph_endpoint(url):
+    ceph = urlparse(url)
+    use_ssl = bool(ceph.scheme == 'https')
+    return {"endpoint": url, "use_ssl": use_ssl}
+
+
+def parse_default_endpoint(url, mode, encryption_mode, dualstack, sig_4):
+    result = {"endpoint": url}
+    config = {}
+    if (mode in ('get', 'getstr') and sig_4) or (mode == "put" and encryption_mode == "aws:kms"):
+        config["signature_version"] = "s3v4"
+    if dualstack:
+        config["s3"] = {"use_dualstack_endpoint": True}
+    if config != {}:
+        result["config"] = Config(**config)
+    return result
+
+
+def get_s3_connection(module, aws_connect_kwargs, location, ceph, endpoint_url, sig_4=False):
+    params = dict(
+        conn_type='client',
+        resource='s3',
+        region=location,
+        **aws_connect_kwargs
+    )
+    if ceph:
+        endpoint_p = parse_ceph_endpoint(endpoint_url)
+    elif is_fakes3(endpoint_url):
+        endpoint_p = parse_fakes3_endpoint(endpoint_url)
+    else:
+        mode = module.params.get("mode")
+        encryption_mode = module.params.get("encryption_mode")
+        dualstack = module.params.get("dualstack")
+        endpoint_p = parse_default_endpoint(endpoint_url, mode, encryption_mode, dualstack, sig_4)
```

Comment:
See also @goneri's comments on https://github.com/ansible-collections/amazon.aws/pull/1156

It's better to pull these out of `module.params` and pass them as explicit arguments.  This code should probably try to be agnostic of the name of the argument being used.